Mathlib/LinearAlgebra/RootSystem/RootPositive.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2024 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+import Mathlib.LinearAlgebra.RootSystem.Defs
+
+/-!
+# Root-positive bilinear forms on root pairings
+
+This file contains basic results on Weyl-invariant inner products for root systems and root data.
+We introduce a Prop-valued mixin class for a root pairing and bilinear form, specifying
+reflection-invariance, symmetry, and strict positivity on all roots.  We show that root-positive
+forms display the same sign behavior as the canonical pairing between roots and coroots.
+
+Root-positive forms show up naturally as the invariant forms for symmetrizable Kac-Moody Lie
+algebras.  In the finite case, the canonical polarization yields a root-positive form that is
+positive semi-definite on weight space and positive-definite on the span of roots.
+
+## Main definitions:
+
+ * `IsRootPositive`: A prop-valued mixin class for root pairings with bilinear forms, specifying
+  the form is symmetric, reflection-invariant, and all roots have strictly positive norm.
+
+## Main results:
+
+* `pairing_pos_iff` and `pairing_zero_iff` : sign relations between `P.pairing` and the form `B`.
+* `coxeter_weight_non_neg` : All pairs of roots have non-negative Coxeter weight.
+* `orthogonal_of_coxeter_weight_zero` : If Coxeter weight vanishes, then the roots are orthogonal.
+
+## TODO
+
+* Invariance under the Weyl group.
+
+-/
+
+noncomputable section
+
+variable {ι R M N : Type*}
+
+namespace RootPairing
+
+variable [LinearOrderedCommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+
+/-- A Prop-valued class for a bilinear form to be compatible with a root system. -/
+class IsRootPositive (P : RootPairing ι R M N) (B : M →ₗ[R] M →ₗ[R] R) : Prop where
+  root_pos : ∀ i : ι, 0 < B (P.root i) (P.root i)
+  symm : ∀ x y : M, B x y = B y x
+  refl_inv : ∀ (i : ι) (x y : M), B (P.reflection i x) (P.reflection i y) = B x y
+
+section
+
+variable (P : RootPairing ι R M N) (B : M →ₗ[R] M →ₗ[R] R) [IsRootPositive P B] (i j : ι) (x y : M)
+
+lemma root_positive : 0 < B (P.root i) (P.root i) := IsRootPositive.root_pos i
+
+-- does this need a better name?
+lemma symmetric : B x y = B y x := IsRootPositive.symm P x y
+
+lemma reflection_invariant (x y : M) : B (P.reflection i x) (P.reflection i y) = B x y :=
+  IsRootPositive.refl_inv i x y
```

Comment:
I'm not absolutely convinced we need these (though I'm happy to be persuaded if you disagree).